### ISSUE TITLE
XInputGetState causes STATUS_ACCESS_VIOLATION

### ISSUE THREAD

**Issue by junglie85** [microsoft/windows-rs#824]
The following snippet causes a `STATUS_ACCESS_VIOLATION` on exit. If I negate the conditional check, no error occurs. I can only assume that there is some kind of memory corruption occurring as a result of a controller actually being connected, but I'm at a loss as to why or how to fix it.
```
for i in 0..XUSER_MAX_COUNT {
    let mut controller_state = unsafe { std::mem::zeroed::<XINPUT_STATE>() };
    if unsafe { XInputGetState(i, &mut controller_state) == ERROR_SUCCESS.0 } {
        println!("{} {:?}", i, controller_state);
    } else {
        // Controller is not connected.
    }
}
```
Outputs:
```
0 XINPUT_STATE { dwPacketNumber: 11, Gamepad: XINPUT_GAMEPAD { wButtons: 0, bLeftTrigger: 0, bRightTrigger: 0, sThumbLX: -32768, sThumbLY: 0, sThumbRX: 0, sThumbRY: 0 } }
error: process didn't exit successfully: `target\debug\horizon.exe` (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)
```
A full [example is available]( https://github.com/ambye85/horizon/tree/xinput-access-violation), which is basically a window with a direct copy of the [Getting Controller State example](https://docs.microsoft.com/en-us/windows/win32/xinput/getting-started-with-xinput#getting-controller-state) from the XInput docs. Is this a problem with the bindings or my use of them?

**Comment by kennykerr** [comment]
A quick check in C++ produces the same results, but as the docs are broken I have no idea whether the API is misbehaving or whether we're holding it wrong. @stevewhims may know what happened to the XInputGetState docs:

https://www.bing.com/search?q=XInputGetState

But as far as I can tell, the Rust bindings are not at fault.

**Comment by stevewhims** [comment]
Adding @karl-bridge-microsoft (and I sent him an email, too) as writer-owner of XInput.

@kennykerr, until the topics can be restored, you can see the content in the repo here: https://cpubwin.visualstudio.com/win32/_git/sdk-api?path=%2Fsdk-api-src%2Fcontent%2Fxinput.

**Comment by kennykerr** [comment]
Thanks Steve, looks like usage is also described here: https://docs.microsoft.com/en-us/windows/win32/xinput/getting-started-with-xinput

Anyway, this looks like an issue outside the scope of the Windows crate.

**Comment by ambye85** [comment]
Thanks for confirming it's not a Windows crate issue. I've been able to get around the issue by loading XInput at runtime and creating a function pointer to `XInputGetState`. In case anyone else comes across this and want's a possible solution:
```
type XInputGetStateFn = extern "system" fn(u32, *mut XINPUT_STATE) -> u32;
extern "system" fn xinput_get_state_stub(_: u32, _: *mut XINPUT_STATE) -> u32 {
    return ERROR_DEVICE_NOT_CONNECTED.0;
}
#[allow(non_upper_case_globals)]
static mut XInputGetState: XInputGetStateFn = xinput_get_state_stub;

unsafe fn load_xinput() {
    let library = LoadLibraryW("xinput1_4.dll");
    if !library.is_null() {
        XInputGetState = std::mem::transmute(GetProcAddress(library, "XInputGetState"));
    }
}
```
Do you know where is best to raise this issue with XInput?

**Comment by kennykerr** [comment]
Hopefully @Karl-Bridge-Microsoft can help.

**Comment by Karl-Bridge-Microsoft** [comment]
@kennykerr, @ambye85 - an engineering fix has resolved the doc issue you reported.

For XBox developer support, please try the [Xbox Developer Forums](https://forums.xboxlive.com/index.html).

**Comment by MaulingMonkey** [comment]
I'd argue this is (partially) a bindings issue - multiple libs provide `XInput*` symbols, and windows-sys chooses the wrong libs.

I recently encountered this when upgrading a non-UWP/UAP desktop app / project from winit 0.26 to winit 0.27.  This brought in `windows-sys` as a new dependency, which in turn caused `XInput*` to link against `XInputUap.dll` instead of `XInput1_4.dll` by introducing `windows.lib` ahead of `xinput.lib` in link order, despite the `Win32_UI_Input_XboxController` feature not being used at all.

`XInputUap.dll` being intended for UWP/UAP apps (where it works fine!) - it's not entirely suprising that it doesn't work properly in a non-UWP/UAP context.  In theory, such UWP-specific imports should be relegated to a `x86_64-uwp-windows-msvc` target, although I realize that's a tier 3 target / awkward to use / most people will use `x86_64-pc-windows-msvc` even for their UWP stuff instead.  XInput 1.4 having been introduced in Windows 8 - at the same time as the UWP stuff - it's also clear that XInputUap is **not** meant to supercede XInput1_4 in a desktop context.  XInputUap also lacks the headset support that 1.4 has, IIRC.

I believe XInput1_4.dll will even work inside a UWP context, although I'll admit I wouldn't bet on it passing the Windows App Cert Kit banned API tests.  UWP apps should probably be considering [`Windows::Gaming::Input::Gamepad`](https://learn.microsoft.com/en-us/uwp/api/windows.gaming.input.gamepad?view=winrt-22621) instead of xinput anyways, though, since it supports XB1 style trigger rumble and otherwise provides a more UWP-y interface?

### Call Stack (XInputUap.dll)
```
      Windows.Gaming.Input.dll!Microsoft::WRL::ActivationFactory<struct Microsoft::WRL::Implements<class Microsoft::WRL::FtmBase,struct Microsoft::WRL::CloakedIid<struct Windows::Gaming::Input::Custom::ICustomGameControllerFactory> >,struct Windows::Gaming::Input::IArcadeStickStatics,struct Windows::Gaming::Input::IArcadeStickStatics2,0>::Release(void)    Unknown
      XInputUap.dll!XInputCore::XInputManager::~XInputManager(void)    Unknown
      XInputUap.dll!_CRT_INIT ()    Unknown
      XInputUap.dll!__DllMainCRTStartup()    Unknown
      ntdll.dll!LdrpCallInitRoutine()    Unknown
      ntdll.dll!LdrShutdownProcess()    Unknown
      ntdll.dll!RtlExitUserProcess()    Unknown
      kernel32.dll!ExitProcessImplementation ()    Unknown
      ...
```
[Microsoft Q&A Thread](https://learn.microsoft.com/en-us/answers/questions/516621/access-violation-in-xinputuapdll-when-closing-the.html)

### Workaround Issues
A partial workaround is to explicitly link xinput.lib inside your executable's build.rs script or main module:
```rust
// build.rs
fn main() {
    println!("cargo:rustc-link-lib=xinput");
}
```
```rust
// src/main.rs
#[link(name = "xinput")] extern {}
```

I say partial as, while several XInput symbols correctly link against XInput1_4.dll when doing this, `XInputEnable` seems to prioritize XInputUap.dll, even if I really attempt to force the issue by bypassing winapi entirely:
```rust
// still links against XInputUap.dll in the presence of `windows-sys` despite `xinput.lib`
// being able to link it against `XInput1_4.dll` ordinal 5 without `window-sys`, and
// XInputGetState linking against `XInput1_4.dll` in the same build!
#[link(name = "xinput")] extern "system" { fn XInputEnable(enable: u32); }
unsafe { XInputEnable(enable.into()) }
```
Oddly, this weird mixed symbol behavior only seems to happen with my Rust builds for reasons I haven't figured out yet.  Even if I explicitly link against the windows-sys 0.36.1 `windows.lib` after `xinput.lib` in a UWP C++ project, all the XInput symbols (including XInputEnable) link against XInput1_4.dll like I'd expect...

**Comment by riverar** [comment]
Agreed. Will re-open and forward this to metadata repo for resolution. We should be linking against XInput 1.4 (XINPUT1_4.DLL). https://learn.microsoft.com/windows/win32/xinput/xinput-versions
